docs/migration-guide-NIO1-to-NIO2.md

```diff
@@ -0,0 +1,70 @@
+- `swift-nio-extras`: `.package(url: "https://github.com/apple/swift-nio.git", from: "1.0.0")`
+- `swift-nio-http2`: `.package(url: "https://github.com/apple/swift-nio-http2.git", from: "1.0.0")`
+- `swift-nio-transport-services`: `.package(url: "https://github.com/apple/swift-nio-transport-services.git", from: "1.0.0")`
+
+After you changes your `Package.swift`, run `swift package update`.
+
+## Step 3: `ctx` to `context`
+
+The NIO community has [correctly](https://swift.org/documentation/api-design-guidelines/#avoid-abbreviations) [pointed](https://github.com/apple/swift-nio/issues/663#issuecomment-442013880) [out](https://github.com/apple/swift-nio/issues/483) that the use of `ctx` for the `ChannelHandlerContext` variables feels alien in Swift. We agree with this and have decided that despite the breakage it will cause, we shall rename all mentions of `ctx` to `context`. Unfortunately, that means that every `ChannelHandler` in existence needs a change. To make it worse, `ChannelHandler`s have default implementations for all events which means if you have a `ChannelHandler` method that still uses `ctx`, your code will still compile but the method will never be called because its first parameter is now named `context`.
+
+Therefore, do yourself a favour and start the conversion from NIO 1 to NIO 2 by running the following command in the root of your package:
+
+```bash
+# if you're developing on macOS / BSD
+find . -name '*.swift' -type f | while read line; do sed -E -i '' 's/([^"])[[:<:]]ctx[[:>:]]([^"])/\1context\2/g' "$line"; done
+```
+
+```bash
+# if you're developing on Linux
+find . -name '*.swift' -type f | while read line; do sed -E -i 's/([^"])\<ctx\>([^"])/\1context\2/g' "$line";  done
+```
+
+What these unwieldy commands do is replace `ctx` if it appears as its own word with `context` in all `.swift` files under the current directory. For the NIO packages themselves that was all that was needed for this huge change.
+
+## Step 4: Use `_NIO1APIShims`
+
+The goal of the `_NIO1APIShims` module (which is _not_ public API and is also untested) is to help you with all the small renames we have done for NIO 2. We recommend to follow these steps to make use of as many automatisms as possible:
+
+1. add `import _NIO1APIShims` to all files where you `import NIO` or any of the other NIO modules like `import NIOHTTP1`. The following command will list you all those files:
+  `find . -name '*.swift' -type f | while read line; do if grep -q "import NIO" "$line"; then echo "$line"; fi; done`
+2. fix all remaining compile errors, if there are any
+3. apply all the compiler-suggested fixits, in Xcode you can
+   1. Press `⌘'` to Jump to Next Issue
+   2. Press `^⎇⌘F` to Fix All Issues in the current file
+   3. Go back to 3.1 until you've gone through all the files and applied all the fixits
+4. fix remaining warnings manually
+5. remove all `import _NIO1APIShims` imports
```

Comment:
just to point this out: Right now (before NIO 2's API "converged" -- which should be very soon), the recommendation is to leave the `import _NIO1APIShims` import in there

Reply:
Just curious, when do you think you will remove this module? 2.1? 3.0?

Reply:
It’s not public API, so it might go any time really. Don’t think we have made up our mind yet.